### PR TITLE
Refactor challenge

### PR DESCRIPTION
**Pull request by pbochis**

### CODE REVIEW

File: app/elements/common/services/challenge-template-service.html [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+			},
+			save: function(challengeTemplate){
+				this.app.startLoading();
+				this.$.postChallengeTemplate.url = util.build('/challenges/templates');
+				this.$.postChallengeTemplate.body = challengeTemplate;
+				this.$.postChallengeTemplate.generateRequest();
+			},
+			getByScheduledChallengeId: function(id){
+				this.app.startLoading();
+				this.$.getSingleChallengeTemplate.url = util.build('/challenges/' + id);
```
Comment (victorbalan) [inline app/elements/common/services/challenge-template-service.html:68]: use the route you will provide on platform

File: app/elements/common/datepicker/date-time-picker.html [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+			},
+			_confirm: function(){
+				var date = this.$.datePicker.date;
+				this.dateTime = new Date(date.getFullYear(), date.getMonth(), date.getDate(), this.$.timePicker.hour, this.$.timePicker.minute, 0, 0);
+				this._setDate(this.dateTime);
+				this.$.dialog.close();
+			},
+			_setDate: function(date){
+//				this.$.textInput.value = date.getFullYear() + '-' + date.getMonth() + 1 + '-' + date.getDate() + ' ' + date.getHours() + ':' + date.getMinutes();
```
Comment (victorbalan) [inline app/elements/common/datepicker/date-time-picker.html]: ??

File: app/elements/common/datepicker/date-time-picker.html [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+			properties: {
+				dateTime: {
+					type: Object,
+					notify: true,
+					reflectToAttribute: true,
+					observer: '_dateTimeSet'
+				},
+				label: {
+					type: String,
+					notify: true
```
Comment (victorbalan) [inline app/elements/common/datepicker/date-time-picker.html:43]: reflectToAttribute

File: app/elements/common/datepicker/date-time-picker.html [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+			<iron-icon prefix icon="receipt"></iron-icon>
+		</paper-input>
+		<paper-dialog id="dialog" class$="{{modalClass}}"modal>
+			<iron-pages selected="{{step}}">
+				<paper-date-picker id="datePicker"></paper-date-picker>
+				<paper-time-picker id="timePicker"></paper-time-picker>
+			</iron-pages>
+			<div class="buttons">
+				<paper-button dialog-dismiss>Cancel</paper-button>
+				<paper-button on-click="_nextStep">Next</paper-button>
```
Comment (victorbalan) [inline app/elements/common/datepicker/date-time-picker.html]: replace Next with Ok/Save if step === 1

File: app/elements/common/services/challenge-service.html [hunk 5]
```diff
@@ -13,26 +13,7 @@
...
-			on-response="_onResponse"
-			on-error="_onError">
-		</iron-ajax>
-		<iron-ajax
-			id="getMultiChallenges"
-			method="GET"
-			handle-as="json"
-			content-type="application/json"
-			last-response="{{challenges}}"
+			body="{{challenge}}"
```
Comment (victorbalan) [inline app/elements/common/services/challenge-service.html]: set body in save function..it`s too brainfucking to have body = challenge and lastresponse = challenge

File: app/elements/company/candidate/invite/invite-candidates.html [hunk 6]
```diff
@@ -76,6 +76,7 @@
 				if (!this.$.form.validate()){
 					return;
 				}
+				console.log(this.candidates.length);
```
Comment (victorbalan) [inline app/elements/company/candidate/invite/invite-candidates.html]: ???

File: app/elements/company/challenge/challenge-create.html [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+	<template>
+		<challenge-service id="challengeService" challenge="{{challenge}}"></challenge-service>
+		<div class="mainContent">
+			<paper-material class="card" elevation="1">
+				<form is="iron-form" id="challengeForm">
+					<paper-input id="name" class="margined" label="Name" value="{{challenge.name}}"></paper-input>
+					<paper-input id="canonicalName" class="margined" label="Canonical name*" value="{{challenge.canonicalName}}" required></paper-input>
+					<paper-checkbox id="inviteOnly" value="{{challenge.inviteOnly}}" checked>Invite only</paper-checkbox>
+					<!--todo inspect why the time picker acts super weird-->
+					<date-time-picker id="startDate" label="Fixed start date" date-time="{{challenge.startDate}}"></date-time-picker>
```
Comment (victorbalan) [inline app/elements/company/challenge/challenge-create.html:32]: don`t add it if it acts super weird

File: app/elements/company/challenge/challenge-modal.html [hunk 8]
```diff
@@ -8,104 +8,58 @@
...
-							<option value="{{item.id}}">{{item.name}}</option><br>
-						</template>
-					</paper-menu>
-				</paper-dropdown-menu>
-				<div class="buttonGroup">
-					<paper-button dialog-dismiss>Cancel</paper-button>
-					<paper-button class="primary" raised on-click="saveChallenge">Create
-					</paper-button>
-				</div>
+				<paper-input id="name" class="margined" label="Name" value="{{challengeTemplateToSave.name}}" required></paper-input>
```
Comment (victorbalan) [inline app/elements/company/challenge/challenge-modal.html:39]: it is challenge modal and challengeTemplateToSave?

File: app/elements/company/challenge/challenge-modal.html [hunk 9]
```diff
@@ -8,104 +8,58 @@
...
+
 			</form>
+				<div class="buttonGroup">
+				<paper-button dialog-dismiss>Cancel</paper-button>
+				<paper-button class="primary" raised on-click="_saveChallenge">Create
```
Comment (victorbalan) [inline app/elements/company/challenge/challenge-modal.html]: no _saveChalleng efunction

File: app/elements/company/challenge/challenge-template-card.html [hunk 10]
```diff
@@ -48,43 +48,39 @@
...
+			computeDuration: function (challengeTemplate) {
+				return util.computeDuration(challengeTemplate.duration);
 			},
-			computeDuration: function (challenge) {
-				return util.computeDuration(challenge.duration);
-			},
-			_viewChallengeDetail: function () {
-				page.redirect('/challenges/' + this.challenge.id);
+			_viewChallengeTemplateDetail: function () {
+				page.redirect('/challenges/' + this.challengeTemplate.id);
```
Comment (victorbalan) [inline app/elements/company/challenge/challenge-template-card.html]: /challenges route should be now /challenges/templates

File: app/elements/company/challenge/challenge-template-list.html [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+<link rel="import" href="/bower_components/iron-ajax/iron-ajax.html">
+
+<dom-module id="challenge-template-list">
+	<style>
+		.challenge-template-list {
+			padding: 10px;
+		}
+	</style>
+	<template>
+		<challenge-template-service id="challengeService" challenge-templates="{{challengeTemplates}}"></challenge-template-service>
```
Comment (victorbalan) [inline app/elements/company/challenge/challenge-template-list.html]: rename to challengeTemplateService

File: app/elements/company/user/user-detail.html [hunk 12]
```diff
@@ -83,14 +83,15 @@
 					</div>
 					<div id="topRight">
 						<div id="skillPolygonWrapper">
+							<skill-polygon id="skillPolygon" skills="{{mockedSkills}}"></skill-polygon>
 						</div>
-						<template is="dom-if" if="{{!hasCompletedChallenges}}">
+						<template is="dom-if" if="{{!hasCompletedchallengeTemplates}}">
```
Comment (victorbalan) [inline app/elements/company/user/user-detail.html]: hasCompletedChallenges is correct..he cannot complete templates

File: app/elements/company/user/user-detail.html [hunk 13]
```diff
@@ -100,10 +101,10 @@ <h3 id="noActivity">User has no activity yet</h3>
 							<paper-button style="float: right;" class="secondary" raised on-click="_loadChallenge">Load more</paper-button>
 						</div>
 					</paper-material>
-					<template is="dom-if" if="{{displayPreviousChallenges}}">
+					<template is="dom-if" if="{{displayPreviouschallengeTemplates}}">
```
Comment (victorbalan) [inline app/elements/company/user/user-detail.html]: same here

File: app/elements/company/user/user-detail.html [hunk 14]
```diff
@@ -100,10 +101,10 @@ <h3 id="noActivity">User has no activity yet</h3>
...
 						<paper-material id="midRight" elevation="1">
-							<h3>Previous challenges</h3>
-							<template is="dom-repeat" items="{{previousChallenges}}">
+							<h3>Previous challenge templates</h3>
```
Comment (victorbalan) [inline app/elements/company/user/user-detail.html]: same here

File: app/scripts/util.js [hunk 15]
```diff
@@ -69,6 +69,21 @@ var util = {
...
 	},
+	formatDate: function(date, ommitTime){
+		var month = Number(date.getMonth()) + 1;
+		var formattedDate = date.getDate() + '-' + month + '-' + date.getFullYear();
+		if (!ommitTime){
+			formattedDate = formattedDate + ' ' + this.getTimeField(date.getHours()) + ':' + this.getTimeField(date.getMinutes());
+		}
+		return formattedDate;
+	},
+	getTimeField: function(timeField){
```
Comment (victorbalan) [inline app/scripts/util.js]: why not create a formatTime function?

File: app/elements/company/user/challenge/user-challenge-detail.html [hunk 16]
```diff
@@ -114,16 +114,7 @@
...
-					this.displaySkillInfo = false;
-				}else{
-					this.displaySkillInfo = true;
-					// TODO: Investigate why we need a timeout of 1 ms.
-					var element = this;
-					window.setTimeout(function(){
-						element.$$('#skillPolygon').buildChart();
-					}, 1);
-				}
+				this.displaySkillInfo = !this.skills;
```
Comment (victorbalan) [inline app/elements/company/user/challenge/user-challenge-detail.html:117]: this does not display the skills if they exist

File: app/scripts/util.js [hunk 17]
```diff
@@ -11,10 +11,10 @@ var util = {
 		return prefix + suffix;
 	},
 	getWSUrl: function(suffix){
-		var prefix = 'wss://ws.cod.uno:8090';
+		var prefix = 'wss://ws.cod.uno:8080';
```
Comment (lorenzleutgeb) [inline app/scripts/util.js:14]: Why did you change this to 8080?
